test/src/test/java/org/corfudb/runtime/object/transactions/OptimisticTransactionContextTest.java

```diff
@@ -842,4 +867,29 @@ public void txnSpansEpochsCanCommitIfPrimarySequencerSpansConsecutiveEpochs() th
         // When t2 wants to commit, the sequencer does not have consecutive epochs, so it should abort.
         t2(this::TXEnd).assertThrows().isInstanceOf(TransactionAbortedException.class);
     }
+
+    /**
+     * Test that the transaction can commit if the first attempt got overwritten and retried.
+     * Basically on overwrite and retry we need to adjust the transaction snapshot otherwise
+     * the transaction will always conflict with itself and get TransactionAbortedException.
+     */
+    @Test
+    public void txnCanCommitIfOverwrittenAndRetry() throws Exception {
+        // Get a new runtime.
+        CorfuRuntime rt = getNewRuntime(getRuntime().getParameters())
+                .parseConfigurationString(getDefaultConfigurationString());
+        rt.connect();
+
+        OptimisticTXBegin();
+
+        put("k1", "v1");
+        put("k2", "v2");
+
+        // Hole fill the address that the transaction is about to commit, to
+        // create an OverwriteException and retry. The retry should succeed.
+        long currentTail = rt.getSequencerView().query().getToken().getSequence();
+        rt.getAddressSpaceView().read(currentTail + 1);
+
+        assertThatCode(this::TXEnd).doesNotThrowAnyException();
```

Comment:
Can you check that the returned address is currentTail + 2 and that the data is there?

Reply:
Done.